base/applications/taskmgr/procpage.c

```diff
@@ -931,3 +936,300 @@ int CALLBACK ProcessPageCompareFunc(LPARAM lParam1, LPARAM lParam2, LPARAM lPara
+    {
+        WCHAR szDrive[3];
+        WCHAR szDevPath[MAX_PATH];
+
+        szDrive[0] = cDrive;
+        szDrive[1] = L':';
+        szDrive[2] = L'\0';
+
+        if (QueryDosDeviceW(szDrive, szDevPath, _countof(szDevPath)) != 0)
+        {
+            size_t len = wcslen(szDevPath);
+            bSuccess = (_wcsnicmp(lpPath, szDevPath, len) == 0);
+
+            if (bSuccess)
+            {
+                StringCchPrintfW(pszPath, dwLength, L"%s%s", szDrive, lpPath + len);
+                break;
+            }
+        }
+    }
+
+    if (bSuccess)
+    {
+        StringCchCopyW(lpPath, dwLength, pszPath);
+    }
+
+    HeapFree(GetProcessHeap(), 0, pszPath);
+
+    return bSuccess;
+}
+
+static DWORD GetProcessExecutablePath(HANDLE hProcess, LPWSTR lpExePath, DWORD dwLength)
+{
+    BYTE StaticBuffer[sizeof(UNICODE_STRING) + (MAX_PATH * sizeof(WCHAR))];
+    PVOID DynamicBuffer = NULL;
+    PUNICODE_STRING ImagePath = NULL;
+    WCHAR *pszExePath = NULL;
+    ULONG SizeNeeded;
+    NTSTATUS Status;
+    DWORD dwRet = 0;
+
+    Status = NtQueryInformationProcess(hProcess,
+                                       ProcessImageFileName,
+                                       StaticBuffer,
+                                       sizeof(StaticBuffer) - sizeof(WCHAR),
+                                       &SizeNeeded);
+
+    if (Status == STATUS_INFO_LENGTH_MISMATCH)
+    {
+        DynamicBuffer = HeapAlloc(GetProcessHeap(), 0, SizeNeeded + sizeof(WCHAR));
+
+        if (!DynamicBuffer)
+        {
+            return 0;
+        }
+
+        Status = NtQueryInformationProcess(hProcess,
+                                           ProcessImageFileName,
+                                           DynamicBuffer,
+                                           SizeNeeded,
+                                           &SizeNeeded);
+
+        ImagePath = (PUNICODE_STRING)DynamicBuffer;
+    }
+    else
+    {
+        ImagePath = (PUNICODE_STRING)StaticBuffer;
+    }
+
+    if (!NT_SUCCESS(Status))
+    {
+        goto Cleanup;
+    }
+
+    pszExePath = HeapAlloc(GetProcessHeap(), 0, ImagePath->Length + sizeof(WCHAR));
+
+    if (!pszExePath)
+    {
+        goto Cleanup;
+    }
+
+    memcpy(pszExePath, ImagePath->Buffer, ImagePath->Length);
+    pszExePath[ImagePath->Length / sizeof(WCHAR)] = UNICODE_NULL;
+
+    if (!DevicePathToDosPath(pszExePath, (ImagePath->Length / sizeof(WCHAR)) + 1))
+    {
+        goto Cleanup;
+    }
+
+    dwRet = wcslen(pszExePath) + 1;
+
+    if (dwLength >= dwRet)
+    {
+        StringCchCopyW(lpExePath, dwLength, pszExePath);
+
+        dwRet -= 1;
+    }
+
+Cleanup:
+
+    if (pszExePath)
+    {
+        HeapFree(GetProcessHeap(), 0, pszExePath);
+    }
+
+    if (DynamicBuffer)
+    {
+        HeapFree(GetProcessHeap(), 0, DynamicBuffer);
+    }
+
+    return dwRet;
+}
+
+static DWORD GetProcessExecutablePathById(DWORD dwProcessId, LPWSTR lpExePath, DWORD dwLength)
+{
+    DWORD dwRet = 0;
```

Comment:
Question for here and the other function above: What does ret == 0 mean? and -1 ? Success or failure?

Reply:
If the function succeeds, the return value is the path length, not including the terminating null character.

If the function fails, the return value is zero.

If the buffer is not large enough, the return value is the required size of the buffer, in characters, including the terminating null character.